simple_functorch.py

```diff
@@ -1277,3 +1305,38 @@ def run_gradvmap(d2: "Batched", d1: "Autograd"):
 
 
 run_gradvmap(d2, d1)
+
+d = Batched(Torch(), length=3)
+
+# Custom vmap
+def f(d, x):
+    return d.mul(x, x)
+
+def f_batch_rule(d, x):
+    # to prove a point
+    return d.add(x, x)
+
+x = label(torch.tensor([1., 2., 3.]))
+
+result = d.custom_vmap(f, f_batch_rule, x)
+assert torch.allclose(result, 2 * x)
+
+# autograd should let custom_vmap pass through
+x = label(torch.tensor([1., 2., 3.]))
+d = Autograd(Torch())
+result = d.custom_vmap(f, f_batch_rule, x)
+assert torch.allclose(result, x * x)
+loss999 = d.sum(result, name='loss999')
+
+grad_x, = d.grad(loss999, [x])
+assert torch.allclose(grad_x, 2 * x)
+
+# autograd should let custom_vmap pass through
+x = label(torch.tensor([[1., 2., 3.], [4., 5., 6.]]))
+d = Autograd(Batched(Torch(), length=2))
+result = d.custom_vmap(f, f_batch_rule, x)
+assert torch.allclose(result, 2 * x)
+loss88 = d.sum(result, name='loss88')
+
+grad_x, = d.grad(loss88, [x])
+assert torch.allclose(grad_x, torch.full_like(x, 2))
```

Comment:
I don't think I agree with this?
I would expect that `Autograd(Batched(Torch(), length=2))` would give `2 * x` as the gradient while `Batched(Autograd(Torch()), length=2)` would give `2`. No?

Reply:
Yes.
I would expect here to be able to see the different between batch of gradients and element-wise gradients.

Reply:
Isn't that question beyond what we're discussing here?
The question is still there without considering any custom_vjp right?

Reply:
Oh I'm just trying to understand the difference between this and custom_vjp. I think the current semantics are analogous to what custom_vjp is doing but reasoning through it is confusing

Reply:
cc @ezyang @Chillee -- this is what we were discussing during the Composability hangout, if you folks have opinions

Reply:
Disregarding custom_vjp for a moment, if we're going off of the argument from the meeting today that this should match the behavior of a normal batch rule, isn't this code right?

Let's say we're calling `d2.unsqueeze(x, dim)` with `d2 = Batched(Autograd(Torch()))`. First it hits the Batched dispatcher, so we get `d2.inner.unsqueeze(x, dim + 1)` so the Autograd dispatcher sees `unsqueeze(x, dim + 1)` (which is the batch rule) and does autograd on _that_ function. Using the same dispatcher stack, we similarly expect autograd runs on the "batch rule"

As a related a note, I always get confused that `Batched(Autograd(Torch()))` is the same as `grad(vmap())` so it might be worth to add the transform implementations if that's not too much work? I think this makes a lot more sense if we're able to say `grad(vmap())` gets the derivative of the custom batch rule rather than remembering to reason about the interpreter stack.

Reply:
I agree that the current Dispatcher for the regular code does that.
But now if we call the `custom_vmap` with this unsqueeze function does it do that as well?

Reply:
> But now if we call the custom_vmap with this unsqueeze function does it do that as well?

Yes?

More explanation:
> I would expect that Autograd(Batched(Torch(), length=2)) would give 2 * x as the gradient

Given that the original function is `d.mul(x, x)`, this would mean that autograd is running on the unbatched function, not the batch rule. If we agree that given this set of Dispatchers should end up with Autograd running on the batched rule, then 2 is the expected gradient because `d.add(x, x)` is the batched rule and the derivative of that is 2